plenum/server/monitor.py

```diff
@@ -51,6 +52,77 @@ def get_avg(metrics: List):
         return median_high(metrics)
 
 
+class MonitorStrategy(ABC):
+    @abstractmethod
+    def add_instance(self):
+        pass
+
+    @abstractmethod
+    def remove_instance(self):
+        pass
+
+    @abstractmethod
+    def reset(self):
+        pass
+
+    @abstractmethod
+    def update_time(self, timestamp: float):
+        pass
+
+    @abstractmethod
+    def request_received(self, id: str):
+        pass
+
+    @abstractmethod
+    def request_ordered(self, id: str, inst_id: int):
+        pass
+
+    @abstractmethod
+    def is_master_degraded(self) -> bool:
+        return False
+
+
+class AccumulatingMonitorStrategy(MonitorStrategy):
+    def __init__(self, instances: int, threshold: int, timeout: float):
+        self._instances = instances
+        self._threshold = threshold
+        self._timeout = timeout
+        self._ordered = defaultdict(lambda: 0)
+        self._timestamp = None
+        self._alert_timestamp = None
+
+    def add_instance(self):
+        self._instances += 1
+
+    def remove_instance(self):
+        self._instances -= 1
+
+    def reset(self):
+        self._alert_timestamp = None
+        self._ordered.clear()
+
+    def update_time(self, timestamp: float):
+        self._timestamp = timestamp
+        master_ordered = self._ordered[0]
+        max_ordered = max(self._ordered[i] for i in range(1, self._instances))
```

Comment:
Should we check for max, or average, or median? 
What if a malicious backup Primary starts ordering much faster than others just by re-ordering the same requests?

Reply:
After more thinking, I would say that taking MAX maybe a good idea. Even if F malicious backups force view change, so that one of them becomes a master Primary and starts slow ordering, 1 valid instance will order faster, and, since we take MAX, it will be noticed.